src/renate/utils/pytorch.py

```diff
@@ -156,3 +156,62 @@ def complementary_indices(num_outputs: int, valid_classes: Set[int]) -> List[int
         valid_classes: A set of integers of valid classes.
     """
     return [class_idx for class_idx in range(num_outputs) if class_idx not in valid_classes]
+
+
+class ConcatRandomSampler(BatchSampler):
+    """Sampler for sampling batches from ConcatDatasets.
+
+    Args:
+        dataset_lengths: The length for the different datasets.
+        batch_sizes: Batch sizes used for specific datasets.
+        complete_dataset_iteration: Provide an index to indicate over which dataset to fully
+            iterate. By default, stops whenever iteration is complete for any dataset.
+        generator (Generator): Generator used in sampling.
+    """
+
+    def __init__(
+        self, dataset_lengths, batch_sizes, complete_dataset_iteration=None, generator=None
+    ) -> None:
+        self.batch_sizes = batch_sizes
+        self.complete_dataset_iteration = complete_dataset_iteration
+        self.subset_samplers = []
+        start_idx = 0
+        num_batches = []
+        for dataset_length, batch_size in zip(dataset_lengths, batch_sizes):
+            end_idx = start_idx + dataset_length
+            self.subset_samplers.append(
+                BatchSampler(
+                    SubsetRandomSampler(list(range(start_idx, end_idx)), generator),
+                    batch_size,
+                    True,
+                )
+            )
+            num_batches.append(dataset_length // batch_size)
+            start_idx = end_idx
+        self.length = (
+            min(num_batches)
+            if complete_dataset_iteration is None
+            else num_batches[self.complete_dataset_iteration]
+        )
+
+    def __iter__(self) -> Iterator[List[int]]:
+        if self.complete_dataset_iteration is None:
+            for samples in zip(*self.subset_samplers):
+                yield [j for i in samples for j in i]
+        else:
+            iterators = [iter(sampler) for sampler in self.subset_samplers]
+            for s in iterators[self.complete_dataset_iteration]:
```

Comment:
Is this optimized? Nested for-loops for each batch seems like a lot.

Reply:
there is no nested loop for each batch. it is a single loop over each iterator. in case 1 this is hidden within `zip` but it also has a loop over each iterator and calls next.